### PR TITLE
Dont crash generate variable

### LINKED ISSUE #45367: CSharpGenerateVariableCodeFixProvider crash -  Destination type must be a TypeDeclarationSyntax
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/1082525/csharpgeneratevariablecodefixprovider-crash-destin.html)._

---
I got a note that it crashed. Here is the provided stack trace. It happened when I clicked on the tip on the left for a red underlined keyword. I had some compiler errors earlier in the file, and it looks like the rest of the file is confused and the brackets might be wrong, making the rest of the file at namespace declaration scope somehow.

System.ArgumentException : Destination type must be a TypeDeclarationSyntax or a CompilationUnitSyntax, but given one is NamespaceDeclarationSyntax.
Parameter name: destination
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.CheckDeclarationNode[TDeclarationNode1,TDeclarationNode2](SyntaxNode destination)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.CSharpCodeGenerationService.AddProperty[TDeclarationNode](TDeclarationNode destination,IPropertySymbol property,CodeGenerationOptions options,IList`1 availableIndices)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.<>c__DisplayClass51_0. <AddPropertyAsync>b__0(SyntaxNode t,CodeGenerationOptions opts,IList`1 ai,CancellationToken ct)
   at async Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.GetEditAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedSolutionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.ComputeOperationsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetPreviewOperationsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.GetPreviewResultAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.<>c__DisplayClass11_0. <GetPreviewAsync>b__0(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunctionAsync[T](<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)

---
#### Visual Studio Feedback System on 6/18/2020, 02:22 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Visual Studio Feedback System on 6/18/2020, 09:17 AM: 

<p>Thank you for sharing your feedback! Our teams prioritize action on product issues with broad customer impact. See details at: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq">https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq</a>. In case you need answers to common questions or need assisted support, be sure to use <a target='_blank' href="https://visualstudio.microsoft.com/vs/support/">https://visualstudio.microsoft.com/vs/support/</a>. We’ll keep you posted on any updates to this feedback.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes #45367

The field and property code generation functions throw if the requirements aren't met, so check for the same requirements before offering them.

### CODE REVIEW

**Review by ryzngard — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
i don't love this PR for two reasons.

1. it's not clear to me how this is solving any particular issue.  i.e. *why* in this test case is there not a type-declaration node to generate into.
2. if we take this approach, we should apply it *uniformly*.  that means having the 'CanAdd' helpers for all types of symbols that can be added.  we should also then use that for *all* consumers of this code gen api.

i.e. i would expect we would see a similar crash for many different features for adding all the different types of symbols.  we're just spot fixing one case instead of fixing all of htem.

**Review by CyrusNajmabadi — CHANGES_REQUESTED** [review]
let's chat a little bit about this

File: src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb [hunk 1]
```diff
@@ -1187,8 +1187,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
 
         Private ReadOnly Property INamedTypeSymbol_IsImplicitClass As Boolean Implements INamedTypeSymbol.IsImplicitClass
             Get
-                ' TODO (tomat):
```
Comment (davidwengier) [inline src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:1190]: @tmat This _seemed_ like a straight forward enough change. Any non-obvious implications?

Comment (davidwengier) [reply src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:1190]: @dotnet/roslyn-compiler PTAL

Comment (tmat) [reply src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:1190]: LGTM

**Review by cston — APPROVED** [review]
Compiler change LGTM.

**Review by RikkiGibson — APPROVED** [review]
Signing off on compiler changes

**Review by AlekseyTs — APPROVED** [review]
VB compiler change LGTM (iteration 1)

**Comment by davidwengier** [comment]
@CyrusNajmabadi can you take another look?

**Comment by CyrusNajmabadi** [comment]
yup/

File: src/EditorFeatures/CSharpTest/GenerateVariable/GenerateVariableTests.cs [hunk 2]
```diff
@@ -9310,6 +9310,28 @@ void Goo()
...
+            await TestExactActionSetOfferedAsync(
+@"using System;
+
+namespace ConsoleApp5
+{
+    class MyException: Exception
+    
+    internal MyException(int error, int offset, string message) : base(message)
+    {
+        [|Error|] = error;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/GenerateVariable/GenerateVariableTests.cs:9329]: so this PR lgtm.  can you try a couple of cases beyond this to see if any other features crash (like gen-method, etc.).

File: src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.cs [hunk 3]
```diff
@@ -130,6 +134,11 @@ private static void AddFieldCodeActions(ArrayBuilder<CodeAction> result, Semanti
         {
             if (state.TypeToGenerateIn.TypeKind != TypeKind.Interface)
             {
+                if (!state.CanGeneratePropertyOrField())
+                {
+                    return;
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.cs:140]: should this be checked prior to even calling AddFieldCodeActions and AddPropertyCodeActions ?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.cs:140]: if this does stay in this helper, i woudl say this should be the absolute first check in teh method.  after all, if someone calls AddField, the first thing we should do is check CanGenField and bail if not.

**Review by CyrusNajmabadi — APPROVED** [review]
approving... though i think this can be rejiggered a bit for cleanliness.  also, a little smoke testing of this area woudl be helpful.

**Comment by davidwengier** [comment]
Thanks. Generate method works fine, will try a few others, but going merge this for 16.8 anyway.

**Pull request merged**